### PR TITLE
 [Workflow] Introduce concept of SupportStrategyInterface

### PR DESCRIPTION
**Pull request by lyrixx**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | yes
| Tests pass?   | yes
| Fixed tickets | #20751
| License       | MIT
| Doc PR        | -

### CODE REVIEW

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php [hunk 1]
```diff
@@ -293,7 +292,12 @@ private function addWorkflowSection(ArrayNodeDefinition $rootNode)
                                     ->end()
                                 ->end()
                             ->end()
-                            ->scalarNode('initial_place')->defaultNull()->end()
+                            ->scalarNode('support_strategy')
+                                ->cannotBeEmpty()
```
Comment (xabbuh) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php]: This looks wrong to me in case you are using the `supports` option.

Comment (lyrixx) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php]: 1/ cannotBeEmpty != required
2/ I have added tests on it ;)

Comment (xabbuh) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php]: Oops, you're right.

Comment (stof) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php]: @lyrixx this forbids resetting it to null in a subsequent config file though

Comment (lyrixx) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php]: I saw your comment in the previous PR. But I don't see a valid use case here.

File: src/Symfony/Bundle/FrameworkBundle/Resources/config/schema/symfony-1.0.xsd [hunk 2]
```diff
@@ -232,7 +232,8 @@
     <xsd:complexType name="workflow">
         <xsd:sequence>
             <xsd:element name="marking-store" type="marking_store" />
-            <xsd:element name="support" type="xsd:string" minOccurs="1" maxOccurs="unbounded" />
+            <xsd:element name="support" type="xsd:string" minOccurs="0" maxOccurs="unbounded" />
+            <xsd:element name="support-strategy" type="xsd:string" minOccurs="0" maxOccurs="1" />
```
Comment (stof) [inline src/Symfony/Bundle/FrameworkBundle/Resources/config/schema/symfony-1.0.xsd]: this should be an attribute, not a child element, as having several ones does not make any sense

Comment (lyrixx) [inline src/Symfony/Bundle/FrameworkBundle/Resources/config/schema/symfony-1.0.xsd]: `minOccurs="0" maxOccurs="1"` it's already bounded. so it's not possible to have many element.
I did not set an attribute to be consistent with `support`.

Comment (stof) [inline src/Symfony/Bundle/FrameworkBundle/Resources/config/schema/symfony-1.0.xsd]: but this is inconsistent with **all** configuration we have in Symfony. We use a child element only for complex values (array nodes in the Configuration class)

Comment (lyrixx) [inline src/Symfony/Bundle/FrameworkBundle/Resources/config/schema/symfony-1.0.xsd]: ok, I will update it ;)

Comment (lyrixx) [inline src/Symfony/Bundle/FrameworkBundle/Resources/config/schema/symfony-1.0.xsd]: Fixed.

**Comment by lyrixx** [comment]
:+1: (for the intial part)

**Comment by stof** [comment]
👍

File: src/Symfony/Component/Workflow/CHANGELOG.md [hunk 3]
```diff
@@ -1,2 +1,8 @@
 CHANGELOG
 =========
+
+3.3.0
+-----
+
+ * Deprecated class name support in `WorkflowRegistry::add()` as second parameter.
+   Wrap the class name in an instance of ClassInstanceSupportStrategy instead.
```
Comment (fabpot) [inline src/Symfony/Component/Workflow/CHANGELOG.md]: You should probably mention the new `support_strategy` option as well (not sure if it makes sense to do it here as it's related to the config in FrameworkBundle).

Comment (fabpot) [inline src/Symfony/Component/Workflow/CHANGELOG.md]: Actually, this should be added to the docs instead.

**Comment by fabpot** [comment]
Thank you @lyrixx.

**Comment by fabpot** [comment]
Thank you @andesk for the initial work on this.

**Comment by andesk** [comment]
👍 

Thank you for taking over, @lyrixx :)

**Pull request merged**